src/sage/schemes/generic/morphism.py

```diff
@@ -337,13 +342,43 @@ def _repr_(self):
             s += "\n  To:   %s" % self._codomain
         d = self._repr_defn()
         if d != '':
-            s += "\n  Defn: %s" % ('\n        '.join(self._repr_defn().split('\n')))
+            s += "\n  Defn: %s" % '\n        '.join(d.split('\n'))
         return s
 
+    def _add_(self, other):
+        r"""
```

Comment:
I think this `_add_` method should only be provided when the underlying scheme is in the category of additive magmas. 
```
sage: E = EllipticCurve(j=42)
sage: E.categories()
[Category of abelian varieties over Rational Field,
 Category of commutative additive groups,
 Category of additive groups,
 Category of additive inverse additive unital additive magmas,
 Category of commutative additive monoids,
 Category of additive monoids,
 Category of additive unital additive magmas,
 Category of commutative additive semigroups,
 Category of additive commutative additive magmas,
 Category of additive semigroups,
 Category of additive magmas,
 Category of schemes over Rational Field,
 Category of schemes,
 Category of sets,
 Category of sets with partial maps,
 Category of objects]
```

Reply:
The problem: For Jacobians of HyperellipticCurve, nothing is declared. What laws does the addition of points of it satisfy?
```
sage: R.<x> = QQ[]
            sage: J = HyperellipticCurve(x^5 + x + 1).jacobian()
            sage: phi = End(J).identity()
sage: EJ = End(J)
sage: EJ.categories()
[Category of endsets of schemes over Rational Field,
 Category of endsets,
 Category of monoids,
 Category of semigroups,
 Category of unital magmas,
 Category of magmas,
 Category of homsets of schemes over Rational Field,
 Category of homsets,
 Category of sets,
 Category of sets with partial maps,
 Category of objects]
```

Reply:
It has an abelian group structure - it's basically a $\mathbb{Z}$-module (with some group quotients).

Reply:
I will look into how to add this categories structure, I'm not too familiar with it.

Reply:
I have now made Jacobians (over fields) abelian varieties:
```diff
# a.patch
# this is ok since Jacobian constructor only accepts curves over fields
-         Scheme.__init__(self, C.base_scheme())
+         Scheme.__init__(self, C.base_scheme(), category=AbelianVarieties(C.base_ring()))
```
Then
```sage
sage: EllipticCurve(QQ, [0, 1]).categories() == HyperellipticCurve(polygen(QQ)^5 + 1).jacobian().categories()
True
```

Reply:
I have fixed the jacobian issue so that (nice) jacobians have an abelian variety structure.
```sage
sage: EllipticCurve(j=42).categories()
[Category of abelian varieties over Rational Field,
 Category of commutative additive groups,
 Category of additive groups,
 Category of additive inverse additive unital additive magmas,
 Category of commutative additive monoids,
 Category of additive monoids,
 Category of additive unital additive magmas,
 Category of commutative additive semigroups,
 Category of additive commutative additive magmas,
 Category of additive semigroups,
 Category of additive magmas,
 Category of schemes over Rational Field,
 Category of schemes,
 Category of sets,
 Category of sets with partial maps,
 Category of objects]
sage: Jacobian_generic(HyperellipticCurve(x^5 + 1)).categories() == EllipticCurve(j=42).categories()
True
```

You mentioned that
> I think this `_add_` method should only be provided when the underlying scheme is in the category of additive magmas.

How should I implement this? Should I check if `self in AdditiveMagmas() and other in AdditiveMagmas()`?

Reply:
I think you can just provide it via the category, as `AbelianVarieties.MorphismMethods._add_` etc.